File: 007/Python/problem007.py
def nth_prime(n):
    if n == 1:
        return 2
    else:
        return nth_prime_after(n-1, 3, 1 * 2)
        
def nth_prime_after(n, p, product_of_primes):
    if n == 1:
        return p
    else:
        return nth_prime_after(n-1, next_prime(p+2, product_of_primes), product_of_primes * p)

def next_prime(n, product_of_primes):
    if is_prime(n, product_of_primes):
        return n
    else:
        return next_prime(n+2, product_of_primes)

def is_prime(n, product_of_primes):
    return gcd(product_of_primes, n) == 1

def gcd(a, b):
    if b == 0:
        return a
    else:
        return gcd(b, a % b)

File: 007/Ruby/problem007.rb
class Integer

  def nth_prime
    if self == 1
      2
    else
      (self-1).nth_prime_after(3, 1 * 2)
    end
  end

  def nth_prime_after(p, product_of_primes)
    n = self
    while n > 1
      old_p = p
      p = (p+2).next_prime(product_of_primes)
      product_of_primes *= old_p
      n = n - 1
    end
    p
  end

  def next_prime(product_of_primes)
    n = self
    while not n.is_prime(product_of_primes) do
      n = n + 2
    end
    n
  end
  
  def is_prime(product_of_primes)
    gcd(product_of_primes) == 1
  end

  def gcd(b)
    a = self
    while b != 0
      prev_a = a
      a = b
      b = prev_a.remainder(b)
    end
    a
  end

end
